From: Felix Fietkau <nbd@nbd.name>
Date: Sun, 24 Jan 2016 01:03:51 +0100
Subject: [PATCH] MIPS: fix cache flushing for highmem pages

Most cache flush ops were no-op for highmem pages. This led to nasty
segfaults and (in the case of page_address(page) == NULL) kernel
crashes.

Fix this by always flushing highmem pages using kmap/kunmap_atomic
around the actual cache flush. This might be a bit inefficient, but at
least it's stable.

Signed-off-by: Felix Fietkau <nbd@nbd.name>
---

--- a/arch/mips/mm/cache.c
+++ b/arch/mips/mm/cache.c
@@ -115,6 +115,13 @@ void __flush_anon_page(struct page *page
 {
 	unsigned long addr = (unsigned long) page_address(page);
 
+	if (PageHighMem(page)) {
+		addr = (unsigned long)kmap_atomic(page);
+		flush_data_cache_page(addr);
+		__kunmap_atomic((void *)addr);
+		return;
+	}
+
 	if (pages_do_alias(addr, vmaddr)) {
 		if (page_mapcount(page) && !Page_dcache_dirty(page)) {
 			void *kaddr;
